fvm/evm/emulator/database/database.go

```diff
@@ -0,0 +1,383 @@
+//
+// this method can be merged as part of batcher
+func (db *Database) Commit() error {
+	err := db.storage.Commit()
+	if err != nil {
+		return types.NewFatalError(err)
+	}
+	return nil
+}
+
+// Close commits db changes
+func (db *Database) Close() error {
+	return db.Commit()
+}
+
+// NewBatch creates a write-only key-value store that buffers changes to its host
+// database until a final write is called.
+func (db *Database) NewBatch() gethDB.Batch {
+	return &batch{
+		db: db,
+	}
+}
+
+// NewBatchWithSize creates a write-only database batch with pre-allocated buffer.
+func (db *Database) NewBatchWithSize(size int) gethDB.Batch {
+	return &batch{
+		db:     db,
+		writes: make([]keyvalue, 0, size),
+	}
+}
+
+// NewIterator is not supported in this database
+// if needed in the future we could implement it using atree iterators
+func (db *Database) NewIterator(prefix []byte, start []byte) gethDB.Iterator {
+	panic(types.ErrNotImplemented)
+}
+
+// NewSnapshot is not supported
+func (db *Database) NewSnapshot() (gethDB.Snapshot, error) {
+	return nil, types.ErrNotImplemented
+}
+
+// Stat method is not supported
+func (db *Database) Stat(property string) (string, error) {
+	return "", types.ErrNotImplemented
+}
+
+// Compact is a no op
+func (db *Database) Compact(start []byte, limit []byte) error {
+	return nil
+}
+
+// Len returns the number of entries currently present in the memory database.
+//
+// Note, this method is only used for testing (i.e. not public in general) and
+// does not have explicit checks for closed-ness to allow simpler testing code.
+func (db *Database) Len() int {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	return db.storage.Count()
+}
+
+// keyvalue is a key-value tuple tagged with a deletion field to allow creating
+// memory-database write batches.
+type keyvalue struct {
+	key    []byte
+	value  []byte
+	delete bool
+}
+
+// batch is a write-only memory batch that commits changes to its host
+// database when Write is called. A batch cannot be used concurrently.
+type batch struct {
+	db     *Database
+	writes []keyvalue
+	size   int
+}
+
+// Put inserts the given value into the batch for later committing.
+func (b *batch) Put(key, value []byte) error {
+	return b.set(key, value, false)
+}
+
+// Delete inserts the a key removal into the batch for later committing.
+func (b *batch) Delete(key []byte) error {
+	return b.set(key, nil, true)
+}
+
+func (b *batch) set(key []byte, value []byte, delete bool) error {
+	b.writes = append(b.writes, keyvalue{gethCommon.CopyBytes(key), gethCommon.CopyBytes(value), delete})
+	b.size += len(key) + len(value)
+	return nil
+}
+
+// ValueSize retrieves the amount of data queued up for writing.
+func (b *batch) ValueSize() int {
+	return b.size
+}
+
+// Write flushes any accumulated data to the memory database.
+func (b *batch) Write() error {
+	return b.db.ApplyBatch(b)
+}
+
+// Reset resets the batch for reuse.
+func (b *batch) Reset() {
+	b.writes = b.writes[:0]
```

Comment:
nit: maybe reset `writes` elements to release memory if value is large.

Reply:
I think we don't use this functionality in our code yet, so I added as a TODO for future improvements